File: modules/format/src/main/scala/be/wegenenverkeer/atomium/format/Entry.scala
package be.wegenenverkeer.atomium.format

import java.time.OffsetDateTime

/**
 * Representation of an entry in an Atom feed. The entry acts as a container for metadata and data associated with the
 * entry.
 *
 * @tparam T the type of entry
 */
trait Entry[+T] {

  /** the id of the entry */
  def id: String

  /** the content in the entry */
  def updated: OffsetDateTime

  /** when the entry was last updated */
  def content: Content[T]

  /** links associated with this entry */
  def links: List[Link]
}

File: modules/format-java/src/main/java/be/wegenenverkeer/atomium/format/Entry.java
package be.wegenenverkeer.atomium.format;

import be.wegenenverkeer.atomium.format.pub.AtomPubEntry;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;
import java.time.OffsetDateTime;
import java.util.List;


@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "_type", defaultImpl = AtomEntry.class)
@JsonSubTypes({
        @JsonSubTypes.Type(value = AtomPubEntry.class, name = "atom-pub"),
        @JsonSubTypes.Type(value = AtomEntry.class, name = "atom")})
@XmlJavaTypeAdapter(Adapters.AtomEntryAdapter.class)
public abstract class Entry<T> {

    public abstract String getId();

    public abstract OffsetDateTime getUpdated();

    public abstract Content<T> getContent();

    public abstract List<Link> getLinks();

}